File: lc_soln/0037_sudoku_solver/solution.py
from typing import List

DIM = 9
SDIM = 3
NUMS = "123456789"


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """

        def box_idx(i, j):
            return (i // SDIM) * SDIM + j // SDIM

        rows = [set() for _ in range(DIM)]
        cols = [set() for _ in range(DIM)]
        boxs = [set() for _ in range(DIM)]
        blanks = []

        for i in range(DIM):
            for j in range(DIM):
                c = board[i][j]
                if c == ".":
                    blanks.append((i, j))
                else:
                    rows[i].add(c)
                    cols[j].add(c)
                    boxs[box_idx(i, j)].add(c)

        def solve(idx):
            if idx == len(blanks):
                return True
            i, j = blanks[idx]
            k = box_idx(i, j)
            for c in NUMS:
                if c not in rows[i] and c not in cols[j] and c not in boxs[k]:
                    board[i][j] = c
                    rows[i].add(c)
                    cols[j].add(c)
                    boxs[k].add(c)
                    if solve(idx + 1):
                        return True
                    board[i][j] = "."
                    rows[i].remove(c)
                    cols[j].remove(c)
                    boxs[k].remove(c)
            return False

        solve(0)

File: lc_soln/0037_sudoku_solver/solution.cpp
#include <string>
#include <utility>
#include <vector>

using namespace std;

class Solution {
    static constexpr int kDim = 9;
    static constexpr int kSDim = 3;
    static constexpr string kNums = "123456789";

    vector<int> rows;
    vector<int> cols;
    vector<int> boxs;
    vector<pair<int, int>> blanks;

public:
    void solveSudoku(vector<vector<char>>& board) {
        rows.assign(kDim, 0);
        cols.assign(kDim, 0);
        boxs.assign(kDim, 0);
        for (int i = 0; i < kDim; ++i) {
            for (int j = 0; j < kDim; ++j) {
                char c = board[i][j];
                if (c == '.') {
                    blanks.push_back({i, j});
                } else {
                    int offset = c - '1';
                    rows[i] |= 1 << offset;
                    cols[j] |= 1 << offset;
                    boxs[boxIdx(i, j)] |= 1 << offset;
                }
            }
        }
        solve(board, 0);
    }

    inline int boxIdx(const int i, const int j) const {
        return (i / kSDim) * kSDim + j / kSDim;
    }

    inline bool isValid(const int i, const int j, const int k,
                        const int offset) const {
        return !(((rows[i] >> offset) & 1) | ((cols[j] >> offset) & 1) |
                  ((boxs[k] >> offset) & 1));
    }

    bool solve(vector<vector<char>>& board, const int idx) {
        if (idx == blanks.size()) {
            return true;
        }
        auto [i, j] = blanks[idx];
        int k = boxIdx(i, j);
        for (const char& num : kNums) {
            int offset = num - '1';
            if (isValid(i, j, k, offset)) {
                board[i][j] = num;
                rows[i] |= 1 << offset;
                cols[j] |= 1 << offset;
                boxs[k] |= 1 << offset;
                if (solve(board, idx + 1)) {
                    return true;
                }
                board[i][j] = '.';
                rows[i] &= ~(1 << offset);
                cols[j] &= ~(1 << offset);
                boxs[k] &= ~(1 << offset);
            }
        }
        return false;
    }
};

